tests/p2p/test_sync.py

```diff
@@ -305,6 +312,145 @@ def test_downloader(self):
+            and it is not retried
+          - peer_X re-connects, sync starts and tx_A and tx_B are added to the downloader
+          - peer_Y re-connects, sync starts and tx_A and tx_B are added to the downloader
+          - download of tx_B timeouts, since now there are peers to download it from, it isn't removed and a retry is
+            triggered
+          - tx_B is downloaded, and now it will be processed before tx_A
+          - tx_A is eventually downloaded (or not, this doesn't have to happen)
+          - tx_B is processed, but it will fail because tx_A has not been added yet
+        """
+        self._downloader_bug_setup()
+
+        # disconnect and wait for the download of tx_A to timeout but not yet the download of tx_B
+        self.conn1.disconnect(Failure(Exception('testing')))
+        self.conn2.disconnect(Failure(Exception('testing')))
+        self.clock.advance(settings.GET_DATA_TIMEOUT - 10.0)
+
+        # reconnect peer_X and peer_Y
+        self.conn1 = FakeConnection(self.manager_bug, self.manager1)
+        self.conn2 = FakeConnection(self.manager_bug, self.manager2)
+
+        # proceed as normal until both peers are back to the connections list
+        for _ in range(50):
+            self.conn1.run_one_step()
+            self.conn2.run_one_step()
+            self.clock.advance(0.1)
+            details_A = self.downloader.pending_transactions.get(self.tx_A.hash)
+            details_B = self.downloader.pending_transactions.get(self.tx_B.hash)
+            details_A_has_conns = details_A is not None and len(details_A.connections) >= 2
+            details_B_has_conns = details_B is not None and len(details_B.connections) >= 2
+            if details_A_has_conns and details_B_has_conns:
+                break
+        else:
+            self.fail('expected to break out of loop')
+
+        # wait for the download of B to be retried
+        self.clock.advance(11.0)
+
+        # this situation should cause the bug before the fix
+        # just advancing the connection a little bit should be enough to finish syncing without the bug
+        for _ in range(20):
+            self.conn2.run_one_step()
+            # self.clock.advance(0.1)
+
+        # if the fix is applied, we would see tx_A in storage by this point
+        self.assertTrue(self.manager_bug.tx_storage.transaction_exists(self.tx_A.hash))
+
+    def test_downloader_disconnect(self):
+        """ This is related to test_downloader_retry_reorder, but it basically tests the change in behavior instead.
+
+        When a peer disconnects it should be immediately remvoed from the tx-detail's connections list.
+        """
+        self._downloader_bug_setup()
+
+        # disconnect and wait for the download of tx_A to timeout but not yet the download of tx_B
+        self.conn1.disconnect(Failure(Exception('testing')))
+        self.conn2.disconnect(Failure(Exception('testing')))
+
```

Comment:
Should we assert that both conn are removed from the tx details connection list?

Reply:
Yeah, you're right. I think I removed it by mistake when refactoring the setup.

Reply:
Fixed.